src/hotspot/share/opto/escape.hpp

```diff
@@ -442,6 +442,10 @@ class ConnectionGraph: public ArenaObj {
         NOT_PRODUCT(trace_es_update_helper(ptn, esc, true, reason));
         ptn->set_fields_escape_state(esc);
       }
+
+      if (esc != PointsToNode::NoEscape) {
```

Comment:
Why is this change needed, i.e., why weren't non-NoEscape objects already set to be non-scalar-replaceable?

Reply:
By default every "PointsToNode" and consequently "JavaObjectNode" is scalar replaceable (see PointsToNode constructor in escape.hpp). In some cases the code only set the JavaObjectNode to ArgEscape/GlobalEscape but doesn't set it to NSR (just search escape.cpp for set_escape_state and you'll find some examples). I don't think that that's ideal and I think making the object NSR as soon as it escape's is a good idea. 

If necessary I can try to revert this change and/or find a case where it's absence causes error.